Removing the use of std::function from tensor.

Neel-Shah-29 opened the pull request:
This fixes #150
 Removing the use of std::function from tensor.

amitsingh19975 commented:
Please, fix all the functions in that file. You need to do the same for all the functions.

Neel-Shah-29 commented:
> Please, fix all the functions in that file. You need to do the same for all the functions. And at the end, please, also squash all the commits into a single one.
Yes sure will do for other functions as well and squash the commits.
I had just a question to be asked that the 
```std::function<void(SizeType r, PointerOut c, PointerIn a)> lambda;```
``` lambda = [&lambda, na, wc, wa, pi](SizeType r, PointerOut c, PointerIn a)```
needs to be changed to 
```auto lambda = [na, wc, wa, pi](auto const& self, SizeType r, PointerOut c, PointerIn a)```

amitsingh19975 commented:
> Please, fix all the functions in that file. You need to do the same for all the functions. And at the end, please, also squash all the commits into a single one.
> Yes sure will do for other functions as well and squash the commits.
> I had just a question to be asked that the
> `std::function<void(SizeType r, PointerOut c, PointerIn a)> lambda;`
> ` lambda = [&lambda, na, wc, wa, pi](SizeType r, PointerOut c, PointerIn a)`
> needs to be changed to
> `auto lambda = [na, wc, wa, pi](auto const& self, SizeType r, PointerOut c, PointerIn a)`

Yes.

Neel-Shah-29 commented:
I have squashed all commits to single one and changed all functions according to my understanding . Please have a look at it. Any mistakes if exists in the pull request will be resolved as soon as possible , Just notify me here. Thanks !

Neel-Shah-29 commented:
```
std::function<void(SizeType r, PointerOut c, PointerIn a)> lambda;  
lambda = [&lambda, n, wc, wa](SizeType r, PointerOut c, PointerIn a)
```
this statement has void return type so here in lambda the trailing return type is to be mentioned as void and above statement can be replaced by below one right?

```
auto lambda = [n, wc, wa](auto const& self, SizeType r, PointerOut c, PointerIn a)->void
```

amitsingh19975 commented:
> ```
> std::function<void(SizeType r, PointerOut c, PointerIn a)> lambda;  
> lambda = [&lambda, n, wc, wa](SizeType r, PointerOut c, PointerIn a)
> ```
> 
> 
>      
>   
> 
> this statement has void return type so here in lambda the trailing return type is to be mentioned as void and above statement can be replaced by below one right?
> 
> 
> ```
> auto lambda = [n, wc, wa](auto const& self, SizeType r, PointerOut c, PointerIn a)->void
> ```
